### ISSUE TITLE
hdf_dump fails for LibriSpeechCorpus 

### ISSUE THREAD

**Issue by akshatdewan** [rwth-i6/returnn#143]
When I run `tools/hdf_dump.py` for `LibriSpeechCorpus`, I get the following error
```
  File "/returnn-experiments/2018-asr-attention/librispeech/full-setup-attention/returnn/HDFDataset.py", line 1029, in dump_from_dataset
    line: shape += dataset.get_data_shape(data_key)
    locals:
      shape = <local> [8507]
      dataset = <local> <LibriWipoEUCorpus 'train' epoch=1>
      dataset.get_data_shape = <local> <bound method Dataset.get_data_shape of <LibriWipoEUCorpus 'train' epoch=1>>
      data_key = <local> 'raw'
  File "/returnn-experiments/2018-asr-attention/librispeech/full-setup-attention/returnn/Dataset.py", line 631, in get_data_shape
    line: if self.num_outputs[key][1] <= 1:
    locals:
      self = <local> <LibriWipoEUCorpus 'train' epoch=1>
      self.num_outputs = <local> {'raw': {'shape': (), 'dtype': 'string'}, 'classes': [10323, 1], 'data': [40, 2]}
      key = <local> 'raw'
KeyError: 1

```
And as you can see in `self.num_outputs = <local> {'raw': {'shape': (), 'dtype': 'string'}, 'classes': [10323, 1], 'data': [40, 2]}`  the  value of the `raw` key of `self.num_outputs` is not consistent with the values of `classes` and `data` keys.

**Comment by albertz** [comment]
My comment from #67:

Yes you are right. No, this is not a problem with your config file. This comes from `LibriSpeechCorpus`, which sets it like this:

    self.num_outputs = {
      "data": [self.num_inputs, 2], "classes": [self.targets.num_labels, 1], "raw": {"dtype": "string", "shape": ()}}

I'm not really sure at this point whether it's wrong to do this (i.e. there is lots of code which does not expect `self.num_outputs` to be like this) (i.e. `LibriSpeechCorpus` must be fixed somehow), or whether that `Dataset.get_data_shape` is kind of an exception (i.e. `Dataset.get_data_shape` should be fixed).

As a simple workaround, you maybe can just remove that `"raw"` entry in `self.num_outputs` there. Probably you don't need that. And I think you also need to remove the `"raw"` handling from `LibriSpeechCorpus._collect_single_seq`.

**Comment by akshatdewan** [comment]
Hi, 

I removed "raw" entry from https://github.com/rwth-i6/returnn/blob/master/GeneratingDataset.py#L1989 and https://github.com/rwth-i6/returnn/blob/master/GeneratingDataset.py#L2200 and it runs fine. I get the following:
```
Horovod initialized. Hostname 0d4622c7aab4, pid 28729, rank 0 / size 1, local rank 0 / local size 1.
Returnn hdf_dump starting up.
{(1, 5): {'max_mean_len': 75, 'subdirs': ['train']}}
{'max_mean_len': 75, 'subdirs': ['train']}
<LibriWipoEUCorpus 'train' epoch=1>, epoch 1. Old mean seq len (transcription) is 141.643195, new is 46.569766, requested max is 75.000000. Old num seqs is 34027, new num seqs is 8507.
<LibriWipoEUCorpus 'train' epoch=1>, epoch 1. Old num seqs 34027, new num seqs 8507.
{(1, 5): {'max_mean_len': 75, 'subdirs': ['train']}}
{'max_mean_len': 75, 'subdirs': ['train']}
<LibriWipoEUCorpus 'train' epoch=1>, epoch 1. Old mean seq len (transcription) is 141.643195, new is 46.569766, requested max is 75.000000. Old num seqs is 34027, new num seqs is 8507.
<LibriWipoEUCorpus 'train' epoch=1>, epoch 1. Old num seqs 34027, new num seqs 8507.
Train data:
  input: 40 x 1
  output: {'data': [40, 2], 'classes': [10323, 1]}
  LibriWipoEUCorpus, sequences: 8507, frames: unknown
Dev data:
  LibriWipoEUCorpus, sequences: 3000, frames: unknown
Source dataset: LibriWipoEUCorpus, sequences: 8507, frames: unknown
Creating HDF dataset file data/cased_asr.hdf
Work on epoch: 1
{(1, 5): {'max_mean_len': 75, 'subdirs': ['train']}}
{'max_mean_len': 75, 'subdirs': ['train']}
<LibriWipoEUCorpus 'train' epoch=1>, epoch 1. Old mean seq len (transcription) is 141.643195, new is 46.569766, requested max is 75.000000. Old num seqs is 34027, new num seqs is 8507.
<LibriWipoEUCorpus 'train' epoch=1>, epoch 1. Old num seqs 34027, new num seqs 8507.
Data keys: ['classes', 'data']
Using input data key: data
Using target data key: classes
Collect stats, iterate through all data...
0:00:00 [|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| 99.99% ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| ]Total len of 'classes' is 98.9K, shape [98896], dtype int32
Total len of 'data' is 3.1M, shape [3096838, 40], dtype float32
Set seq tags...
0:00:00 [|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| 99.99% ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| ]Set seq len info...
0:00:00 [|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| 99.99% ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| ]Create arrays in HDF...
Labels for classes: [b'<s>', b'<unk>', b'the'] ...
Labels for data: ['data-class-0', 'data-class-1', 'data-class-2'] ...
Write data...
{(1, 5): {'max_mean_len': 75, 'subdirs': ['train']}}
{'max_mean_len': 75, 'subdirs': ['train']}
<LibriWipoEUCorpus 'train' epoch=1>, epoch 1. Old mean seq len (transcription) is 141.643195, new is 46.569766, requested max is 75.000000. Old num seqs is 34027, new num seqs is 8507.
<LibriWipoEUCorpus 'train' epoch=1>, epoch 1. Old num seqs 34027, new num seqs 8507.
0:00:00 [|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| 99.98% ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| ]All done.
Quitting

```
I have a question though, it seems that it dumped only 8507 sequences. Is this correct? Because there are around 600,000 seqs in my dataset. What should be the correct cmd line arguments while calling `tools/hdf_dump.py`? I mean for `--start_seq`, `--end_seq` and `--epoch`.

**Comment by albertz** [comment]
I guess you use the train config, right? I think you have used both `partition_epoch` (i.e. only selected some partition of the whole data) and `epoch_wise_filter` (which filters out sequences, and I think you have done that quite aggressively, as I can see from your output).
I would copy the train config, remove everything but the dataset, and in the dataset, remove `partition_epoch` and `epoch_wise_filter`.

`hdf_dump.py` is implemented really straight-forward. It just dumps the first epoch (or the epoch which you specify via `--epoch`) into a HDF. So you have to make sure that this covers all the data you want to have. You don't need to specify `--start_seq` or `--end_seq`.

**Comment by akshatdewan** [comment]
Thanks! I removed the `epoch_wise_filter` and `EpochSplit` and restarted the dump and I guess it will take some time. 
```
Horovod initialized. Hostname 0d4622c7aab4, pid 29147, rank 0 / size 1, local rank 0 / local size 1.
Returnn hdf_dump starting up.
Train data:
  input: 40 x 1
  output: {'data': [40, 2], 'classes': [10323, 1]}
  LibriWipoEUCorpus, sequences: 680530, frames: unknown
Dev data:
  LibriWipoEUCorpus, sequences: 3000, frames: unknown
Source dataset: LibriWipoEUCorpus, sequences: 680530, frames: unknown
Creating HDF dataset file data/cased_asr.hdf
Work on epoch: 1
Data keys: ['classes', 'data']
Using input data key: data
Using target data key: classes
Collect stats, iterate through all data...
13:01:28 [                                                                                       0.13%                                                                                       ]
```
So, now, when `hdf_dump.py` finishes doing its magic I have to train using the generated HDF file, where will I have to make the changes? Should I just change the `LibriSpeechCorpus` class in `GenerateDataset.py` to derive from `HDFDataset` and initialize with HDF filenames?

Training with my "older" (with epoch wise filter and epoch splits) training config should work without modification right?

**Comment by albertz** [comment]
In your train config, you use the new dataset, i.e. it should look sth like `{'class': 'HDFDataset', 'files': [...], ...}`. You don't need to change anything in the code. (And as I explained in #67, `HDFDataset` is not a base class, you cannot use it like that.)

Note that `HDFDataset` also supports `partition_epoch`, and you probably want to use that in the same way. Currently it does not support `epoch_wise_filter`. Maybe just try without that; I'm not sure how important it is for your case.

**Comment by akshatdewan** [comment]
Thanks!

**Comment by akshatdewan** [comment]
Since I was dumping hdf files, I am wondering - what is better, dumping to multiple hdfs or single hdf file? Do you have an opinion on it?

**Comment by albertz** [comment]
This might depend on your hardware or file system. I would prefer a single file.

**Comment by akshatdewan** [comment]
I just wanted to report that I do see an improvement in training time with the use of HDFs. 
```
Stats:
  mem_usage:GPU:0: Stats(mean=5.4GB, std_dev=59.6MB, min=4.9GB, max=5.5GB, num_seqs=2036, avg_data_len=1)
pretrain epoch 36, finished after 2036 steps, 0:45:48 elapsed (95.8% computing time)
Stats:
  mem_usage:GPU:0: Stats(mean=5.5GB, std_dev=56.4MB, min=5.0GB, max=5.5GB, num_seqs=2036, avg_data_len=1)
pretrain epoch 36, finished after 2036 steps, 0:45:48 elapsed (96.6% computing time)
Save model under data/exp-cased2_asr_multigpu/model.pretrain.036
pretrain epoch 36 score: ctc 1.6614775236514112 output/output_prob 1.0492188053537213 elapsed: 0:45:48
reusing previous dataset batch order for 'dev' dataset
TF: log_dir: data/exp-cased2_asr_multigpu/dev-36-2019-03-07-09-49-35
pretrain epoch 36 score: ctc 1.6614775236514112 output/output_prob 1.0492188053537213 elapsed: 0:45:48

```
I now intend to run it for a couple of 100 epochs and compare the results with the single GPU training. I shall report the results when I have them. 

Many thanks for your support.

**Comment by albertz** [comment]
I think this was fixed via a7fa79022. Please reopen if there is still some problem.
